File: pkg/model/image.go
package model

import (
	"net/url"
	"time"

	"github.com/pkg/errors"
)

type ImageItemMask int16

const (
	ImageItemIkedam  = ImageItemMask(0x01)
	ImageItemMinidam = ImageItemMask(0x02)
)

type Image struct {
	ID                 string        `json:"id"`
	ImageURL           string        `json:"imageURL" firestore:"-"`
	ThumbnailURL       string        `json:"thumbnailURL" firestore:"-"`
	TagList            []string      `json:"tagList"`
	Description        string        `json:"description"`
	MakingNote         string        `json:"makingNote"`
	ItemMask           ImageItemMask `json:"itemMask"`
	WithIkedam         bool          `json:"withIkedam" firestore:"-"`
	WithMinidam        bool          `json:"withMinidam" firestore:"-"`
	TwitterURL         string        `json:"twitterURL"`
	TweetComment       string        `json:"tweetComment"`
	PublishTime        time.Time     `json:"publishTime"`
	LastManualTagTime  *time.Time    `json:"lastManualTagTime,omitempty"`
	LastMachineTagTime *time.Time    `json:"lastMachineTagTime,omitempty"`
	CreateTime         time.Time     `json:"createTime"`
	UpdateTime         time.Time     `json:"updateTime"`
}

func (image *Image) OnLoadFromFirestore() error {
	if image.TagList == nil {
		image.TagList = []string{}
	}
	image.WithIkedam = ((image.ItemMask & ImageItemIkedam) != 0)
	image.WithMinidam = ((image.ItemMask & ImageItemMinidam) != 0)
	return nil
}

func (image *Image) FillURL(baseURL string) error {
	imageURL, err := url.JoinPath(baseURL, image.ID)
	if err != nil {
		return errors.Wrap(err, "failed to build URL for image")
	}
	thumbnailURL, err := url.JoinPath(baseURL, "thumbnail", image.ID)
	if err != nil {
		return errors.Wrap(err, "failed to build URL for thumbnail")
	}
	image.ImageURL = imageURL
	image.ThumbnailURL = thumbnailURL
	return nil
}

File: client/src/app/model/image.ts
import * as moment from 'moment-timezone';

export interface Image {
  id: string;
  imageURL: string;
  thumbnailURL: string;
  tagList: string[];
  description: string;
  makingNote: string;
  itemMask: number;
  withIkedam: boolean;
  withMinidam: boolean;
  twitterURL: string;
  tweetComment: string;
  publishTime: moment.Moment;
  lastManualTagTime: moment.Moment|undefined;
  lastMachineTagTime: moment.Moment|undefined;
  createTime: moment.Moment;
  updateTime: moment.Moment;
}

export const DefaultImage: Readonly<Image> = {
  id: '',
  imageURL: '',
  thumbnailURL: '',
  tagList: [],
  description: '',
  makingNote: '',
  itemMask: 0,
  withIkedam: false,
  withMinidam: false,
  twitterURL: '',
  tweetComment: '',
  publishTime: moment.invalid(),
  lastManualTagTime: undefined,
  lastMachineTagTime: undefined,
  createTime: moment.invalid(),
  updateTime: moment.invalid(),
};
